crates/biome_js_runtime/src/plugin_api.rs

```diff
@@ -0,0 +1,76 @@
+use std::cell::RefCell;
+use std::rc::Rc;
+
+use boa_engine::module::SyntheticModuleInitializer;
+use boa_engine::object::FunctionObjectBuilder;
+use boa_engine::{Context, JsValue, Module, NativeFunction, js_string};
+
+use biome_analyze::RuleDiagnostic;
+use biome_diagnostics::{Severity, category};
+use biome_text_size::TextRange;
+
+pub(crate) struct JsPluginApi {
+    diagnostics: Rc<RefCell<Vec<RuleDiagnostic>>>,
+}
+
+impl JsPluginApi {
+    pub(crate) fn new() -> Self {
+        Self {
+            diagnostics: Rc::new(RefCell::new(Vec::new())),
+        }
+    }
+
+    pub(crate) fn create_module(&self, context: &mut Context) -> Module {
+        let diagnostics = self.diagnostics.clone();
+        let add_diagnostic = FunctionObjectBuilder::new(context.realm(), unsafe {
+            NativeFunction::from_closure(move |_this, args, _context| {
+                let [JsValue::String(severity), JsValue::String(message)] = args else {
+                    todo!()
+                };
+
+                let severity = match severity.to_std_string_lossy().as_str() {
+                    "fatal" => Severity::Fatal,
+                    "error" => Severity::Error,
+                    "warning" => Severity::Warning,
+                    "information" => Severity::Information,
+                    "hint" => Severity::Hint,
+                    _ => todo!(),
+                };
+
+                let diagnostic = RuleDiagnostic::new(
+                    category!("plugin"),
+                    None::<TextRange>, // TODO: retrieve a span from the AST
+                    message.to_std_string_lossy(),
+                )
+                .with_severity(severity);
+
+                diagnostics.borrow_mut().push(diagnostic);
+
+                Ok(JsValue::undefined())
+            })
+        })
+        .length(2)
+        .name("addDiagnostic")
+        .build();
+
+        // TODO: auto-generate AST classes and insert into the runtime
+        // TODO: more runtime APIs?
+
+        Module::synthetic(
+            &[js_string!("addDiagnostic")],
```

Comment:
For Grit plugins, we call the function `registerDiagnostic`. Maybe good to keep it consistent?

Reply:
I thought `register_diagnostic` on Grit was named because it's declarative, meaning "register that this pattern will add diagnostic when matched". JS plugins are procedural, so I think `add` is much simple and fits in this case. Let me know what do you think, I will change it to `registerDiagnostic` if my assumption is incorrect.

Reply:
Ah no, that wasn’t the intention behind it at least. I think “add” is more something you do to a number or collection, while “register” is something you do with a person or system. If you add something, you’re the one doing it, and the assumption is that you can also subtract or remove it again. If you register something with an entity, you give responsibility for what happens with the thing to that entity.

From that explanation, I can even see how register does indeed make more sense in a declarative context, because indeed you’re handing control of the registration to the plugin runtime. But the intention was more that plugins don’t manage diagnostics themselves; they’re registering them with Biome and Biome will take care of displaying them.